components/wifi/esp8266-driver/ESP8266Interface.cpp

```diff
@@ -351,41 +377,102 @@ int ESP8266Interface::set_channel(uint8_t channel)
 }
 
 
-int ESP8266Interface::disconnect()
+void ESP8266Interface::_disconnect_async()
 {
     _cmutex.lock();
+    _disconnect_retval = _esp.disconnect() ? NSAPI_ERROR_OK : NSAPI_ERROR_DEVICE_ERROR;
+    int timeleft_ms = ESP8266_INTERFACE_CONNECT_TIMEOUT_MS - _conn_timer.read_ms();
+
+    if (_disconnect_retval == NSAPI_ERROR_OK || ((_if_blocking == true) && (timeleft_ms <= 0))) {
+
+        if (timeleft_ms <= 0 && _connect_retval != NSAPI_ERROR_OK) {
+            _disconnect_retval = NSAPI_ERROR_CONNECTION_TIMEOUT;
+        } else {
+            if (_conn_stat != NSAPI_STATUS_DISCONNECTED) {
+                _conn_stat = NSAPI_STATUS_DISCONNECTED;
+            }
+            // In case the status update arrives later inform upper layers manually
+            _disconnect_event_id = 0;
+            _conn_timer.stop();
+            _connect_retval = NSAPI_ERROR_NO_CONNECTION;
+        }
+        // Power down the modem
+        _rst_pin.rst_assert();
+        _if_connected.notify_all();
+        // Power off the modem
+        _pwr_pin.power_off();
+    } else {
+        // Postpone to give other stuff time to run
+        _disconnect_event_id = _global_event_queue->call_in(
+                                   ESP8266_INTERFACE_CONNECT_INTERVAL_MS,
+                                   callback(this, &ESP8266Interface::_disconnect_async));
+        if (!_disconnect_event_id) {
+            MBED_ERROR(
+                MBED_MAKE_ERROR(MBED_MODULE_DRIVER, MBED_ERROR_CODE_ENOMEM), \
+                "ESP8266Interface::_disconnect_async(): unable to add event to queue. Increase \"events.shared-eventsize\"\n");
+        }
+    }
+    _cmutex.unlock();
+    if (_disconnect_event_id == 0) {
+        if (_conn_stat_cb) {
+            _conn_stat_cb(NSAPI_EVENT_CONNECTION_STATUS_CHANGE, _conn_stat);
+        }
+        _software_conn_stat = IFACE_STATUS_DISCONNECTED;
+    }
+}
+
+int ESP8266Interface::disconnect()
+{
+    if (_software_conn_stat == IFACE_STATUS_DISCONNECTING) {
+        return NSAPI_ERROR_BUSY;
+    }
+    if (_software_conn_stat == IFACE_STATUS_DISCONNECTED) {
+        return NSAPI_ERROR_NO_CONNECTION;
+    }
     if (_connect_event_id) {
         _global_event_queue->cancel(_connect_event_id);
         _connect_event_id = 0; // cancel asynchronous connection attempt if one is ongoing
     }
-    _cmutex.unlock();
-    _initialized = false;
 
-    nsapi_error_t status = _conn_status_to_error();
-    if (status == NSAPI_ERROR_NO_CONNECTION) {
-        return NSAPI_ERROR_NO_CONNECTION;
+    if (!_if_blocking) {
```

Comment:
Is there a reason to acquiring the lock so late?

Reply:
Good point. The mutex should protect the queue - we need to move the lock up, before line 432.

Reply:
I think it should be acquired already before first check of `_software_conn_stat` as it is modified from elsewhere. That's at least my gut feeling.

Reply:
Hmm, we haven't been protecting `_conn_stat` with a mutex so far, I think `_software_conn_stat` could be handled similarly?
I was looking into the documentation but I see no clear indication if `ESP8266Interface` or `NetworkInterface` are meant to be thread-safe or not. If so - then we would probably need to protect even further than we do right now and I would suggest to create a separate task and take care of this outside of this PR. 
For example: we'd have to protect the `set_credentials()` and `get_ip_address()`.
I also wonder if those functions should or shouldn't become properly non-blocking in case `set_blocking(false)` gets called? Somehow the [documentation only mentioned](https://github.com/ARMmbed/mbed-os/blob/d8c2c6f97fe419d682cdc11d859a4ce2952ca23d/features/netsocket/NetworkInterface.h#L309) `connect()` and `disconnect()`...
What do you think @VeijoPesonen ?

Reply:
> I was looking into the documentation but I see no clear indication if ESP8266Interface or NetworkInterface are meant to be thread-safe or not. If so - then we would probably need to protect even further than we do right now and I would suggest to create a separate task and take care of this outside of this PR.

A fair point and I would say we don't need to go so far. What I was thinking here was the interaction between `disconnect_async()` and `disconnect()` when same the caller calls `disconnet()` twice in a row.

Reply:
We've agreed to put the mutex lock after the `_softwatre_conn_stat` checks, as the result will be the same (return of `NSAPI_ERROR_BUSY` either due to failed lock or `_software_conn_stat` value check) and we don't protect that variable elsewhere in the code... Is this ok?

Reply:
Sounds good.